pallets/parachain-staking/src/lib.rs

```diff
@@ -389,287 +449,705 @@ pub mod pallet {
-			Ok((false, amount))
-		}
-		/// Return true if in_top after call
-		/// Caller must verify before call that account is a delegator
-		fn increase_delegation(&mut self, delegator: A, more: B) -> bool {
-			let mut in_top = false;
-			for x in &mut self.top_delegations {
-				if x.owner == delegator {
-					x.amount += more;
-					self.total_counted += more;
-					self.total_backing += more;
-					in_top = true;
-					break;
+			let actual_amount = actual_amount_option.ok_or(Error::<T>::DelegationDNE)?;
+			bottom_delegations.total -= actual_amount;
+			// update candidate info
+			self.reset_bottom_data::<T>(&bottom_delegations);
+			self.delegation_count -= 1u32;
+			<BottomDelegations<T>>::insert(candidate, bottom_delegations);
+			Ok(false)
+		}
+		/// Increase delegation amount
+		pub fn increase_delegation<T: Config>(
+			&mut self,
+			candidate: &T::AccountId,
+			delegator: T::AccountId,
+			bond: BalanceOf<T>,
+			more: BalanceOf<T>,
+		) -> Result<bool, DispatchError>
+		where
+			BalanceOf<T>: Into<Balance>,
+		{
+			let lowest_top_eq_highest_bottom =
+				self.lowest_top_delegation_amount == self.highest_bottom_delegation_amount;
+			let bond_geq_lowest_top = bond.into() >= self.lowest_top_delegation_amount;
+			let delegation_dne_err: DispatchError = Error::<T>::DelegationDNE.into();
+			if bond_geq_lowest_top && !lowest_top_eq_highest_bottom {
+				// definitely in top
+				self.increase_top_delegation::<T>(candidate, delegator.clone(), more)
+			} else if bond_geq_lowest_top && lowest_top_eq_highest_bottom {
+				// update top but if error then update bottom (because could be in bottom because
+				// lowest_top_eq_highest_bottom)
+				let result = self.increase_top_delegation::<T>(candidate, delegator.clone(), more);
+				if result == Err(delegation_dne_err) {
+					self.increase_bottom_delegation::<T>(candidate, delegator, bond, more)
+				} else {
+					result
 				}
+			} else {
+				self.increase_bottom_delegation::<T>(candidate, delegator, bond, more)
 			}
-			// if delegator was increased in top delegations
-			if in_top {
-				self.sort_top_delegations();
-				return true;
-			}
-			// else delegator to increase must exist in bottom
-			// >pop requires push later on to reset in case it isn't used
-			let lowest_top = self
-				.top_delegations
-				.pop()
-				.expect("any bottom delegations => must exist max top delegations");
-			let mut move_2_top = false;
-			for x in &mut self.bottom_delegations {
-				if x.owner == delegator {
-					x.amount += more;
-					self.total_backing += more;
-					move_2_top = x.amount > lowest_top.amount;
-					break;
+		}
+		/// Increase top delegation
+		pub fn increase_top_delegation<T: Config>(
+			&mut self,
+			candidate: &T::AccountId,
+			delegator: T::AccountId,
+			more: BalanceOf<T>,
+		) -> Result<bool, DispatchError>
+		where
+			BalanceOf<T>: Into<Balance>,
+		{
+			let mut top_delegations =
+				<TopDelegations<T>>::get(candidate).expect("existence proof TODO");
```

Comment:
Is this TODO still relevant?

Reply:
I'll write the proof, it is simply the CandidateInfo exists => BottomDelegations exists